### PR TITLE
[quant][graphmode][fx] Add top level APIs

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43728 [quant][graphmode][fx] Add support for weight prepack folding
* **#43581 [quant][graphmode][fx] Add top level APIs**
* #43534 [quant][graphmode][fx][fix] enable per channel quantization for functional ops

Summary:
Add similar APIs like eager and graph mode on torchscript
- fuse_fx
- quantize_static_fx (for post training static quantization)
- quantize_dynamic_fx (for post training dynamic quantization)
- prepare_qat_fx (for qat)
- convert_qat_fx (for qat)

There are also other APIs not necessarily needed by common user:
- prepare_static_fx (for post training static quantization)
- convert_static_fx (for post training static quantization)
- prepare_dynamic_fx (for post training dynamic quantization)
- convert_dynamic_fx (for post training dynamic quantization)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23385091](https://our.internmc.facebook.com/intern/diff/D23385091)

### CODE REVIEW

File: torch/quantization/fx/quantize.py [hunk 1]
```diff
@@ -621,15 +613,16 @@ def get_qconfig(module):
...
-    def _prepare(self, model, qconfig_dict, inplace, quant_type):
+    def _prepare(self, model, qconfig_dict, inplace, is_dynamic):
+        assert not inplace, 'inplace prepare is not supported yet'
         input_root = model.root
         if not inplace:
             input_root = copy.deepcopy(input_root)
 
         input_graph = model.graph
-        self.quant_type = quant_type
+        self.is_dynamic = is_dynamic
```
Comment (vkuzo) [inline torch/quantization/fx/quantize.py:623]: why not keep the enum?  it's more readable

File: torch/quantization/quantize_fx.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
+from .fx import Fuser  # noqa: F401
+from .fx import Quantizer  # noqa: F401
+
+def fuse_fx(graph_module, inplace=False):
+    fuser = Fuser()
+    return fuser.fuse(graph_module, inplace)
+
+def _prepare_fx(graph_module, qconfig_dict, inplace, is_dynamic):
+    quantizer = Quantizer()
```
Comment (vkuzo) [inline torch/quantization/quantize_fx.py:9]: do we need a stateful object for this?  Can it be stateless?

File: torch/quantization/quantize_fx.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+def prepare_fx(graph_module, qconfig_dict, inplace=False):
+    """ If graph_module is in training mode, the model will be
+    prepared as a qat model, otherwise, it will be prepared as a
+    model used for post training static quantization.
+
+    Args:
+      graph_module: model from symbolic_tracing (torch.fx)
+      qconfig_dict: tbd
+    """
```
Comment (vkuzo) [inline torch/quantization/quantize_fx.py:22]: thoughts about adding an assert here that `graph_module` is actually a GraphModule?

File: torch/quantization/quantize_fx.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    quantizer = Quantizer()
+    return quantizer.convert(graph_module, inplace, debug, is_dynamic)
+
+def convert_fx(graph_module, inplace=False, debug=False):
+    return _convert_fx(graph_module, inplace, debug, is_dynamic=False)
+
+def convert_dynamic_fx(graph_module, inplace=False, debug=False):
+    return _convert_fx(graph_module, inplace, debug, is_dynamic=True)
+
+def _quantize_fx(model, qconfig_dict, run_fn=None, run_args=None, inplace=False, debug=False, is_dynamic=False):
```
Comment (vkuzo) [inline torch/quantization/quantize_fx.py:38]: can we use the enum instead of `is_dynamic`, it's more readable

File: torch/quantization/quantize_fx.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        run_fn(model, *run_args)
+        # TODO: change inplace to True since the model is already copied in
+        # prepare
+        model = convert_fx(model, False, debug)
+
+    return model
+
+
+def quantize_fx(model, qconfig_dict, run_fn, run_args, inplace=False, debug=False):
+    r"""Quantize the input float symbolically traced GraphModule model with
```
Comment (vkuzo) [inline torch/quantization/quantize_fx.py:57]: > post training static quantization (if the model is in eval mode) or quantization aware training (if the model is in training mode).

This would be easy to mess up or forget, thoughts about either having separate functions for PTQ versus QAT, or forcing users to choose PTQ vs QAT in an argument and enforce that the `model.training` flag is set correctly based on the quantization mode?

File: torch/quantization/fx/quantize.py (hunk 1 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:623]: it looks like we don't need to distinguish between static and qat. they follow the same flows

File: torch/quantization/quantize_fx.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/quantize_fx.py:9]: There are some shared functions, I think it can be further refactored as functions, but this can be done in future prs

File: torch/quantization/quantize_fx.py (hunk 5 above)
Comment (jerryzh168) [reply torch/quantization/quantize_fx.py:57]: maybe we can also use this api only for post training static. i think for qat, most of the time people will be using prepare and convert

File: torch/quantization/quantize_fx.py (hunk 4 above)
Comment (jerryzh168) [reply torch/quantization/quantize_fx.py:38]: replied in the previous comment, it looks like we do not need to distinguish between static and qat

Comment (vkuzo) [reply torch/quantization/quantize_fx.py:38]: `is_dynamic` can mean many things, `QuantType.DYNAMIC` is very explicit.  Can we use the enum?  It's valuable to have a single enum where we can branch between dynamic, static, QAT, even if not every single piece of code uses all the three branches.

File: torch/quantization/quantize_fx.py (hunk 5 above)
Comment (vkuzo) [reply torch/quantization/quantize_fx.py:57]: sure, sounds reasonable

File: torch/quantization/quantize_fx.py (hunk 4 above)
Comment (jerryzh168) [reply torch/quantization/quantize_fx.py:38]: I mean we don't branch based on static or qat, the only thing we are checking in the code is whether it is dynamic quantization or not, so defining three quant types where two of them are not used seems redundant

Comment (jerryzh168) [reply torch/quantization/quantize_fx.py:38]: I can rename is_dynamic to is_dynamic_quantization to make it more explicit

**Review by vkuzo — APPROVED** [review]

File: torch/quantization/quantize_fx.py (hunk 4 above)
Comment (vkuzo) [reply torch/quantization/quantize_fx.py]: @jerryzh168 , I was just reading through the implementation of `torch/quantization/fx/quantize.py`, and there is a branch for QAT logic, see 

```
if input_root.training:
    self._qat_swap_modules(input_root)
```

can we keep the enum we already have, and switch this line to use it?  There are 3 quant types, with 3 branches in the code.  It's also not ideal to use `root.training` silently, it's not intuitive for this to represent "this is QAT".

Comment (jerryzh168) [reply torch/quantization/quantize_fx.py]: we already used training in the implementation of fusion: https://github.com/pytorch/pytorch/blob/master/torch/quantization/fuse_modules.py#L11, I think it's better to keep consistent. actually I think this branch is not needed, since qat_swap_module will have no effect when we call it on eval models.
